cores/esp8266/PolledTimeout.h

```diff
@@ -45,18 +47,86 @@ struct YieldOrSkip
   static void execute() {delay(0);}
 };
 
+template <unsigned long delayMs>
+struct YieldAndDelayMs
+{
+  static void execute() {delay(delayMs);}
+};
+
 } //YieldPolicy
 
+namespace TimePolicy
+{
+
+struct TimeMillis
+{
+  // time policy in milli-seconds based on millis()
+
+  using timeType = decltype(millis());
+  static timeType time() {return millis();}
+
+  // millis to millis
+  static timeType toTimeTypeUnit (const timeType ms) { return ms; }
+  static timeType toUserUnit (const timeType ms) { return ms; }
+
+  // rollover on 32 bits: 49.7 days (::max() is a "never expires"-reserved value)
+  static constexpr timeType timeMax() { return std::numeric_limits<timeType>::max() - 1; }
+};
+
+struct TimeFastCyclesBase
+{
+  // time policy based on ESP.getCycleCount()
+
+  using timeType = decltype(ESP.getCycleCount());
+  static timeType time() {return ESP.getCycleCount();}
+
+  // this particular time measurement is intended to be called very often
+  // (every loop, every yield)
+};
+
+template <const TimeFastCyclesBase::timeType second_th> struct TimeFastCycles: TimeFastCyclesBase
+{
+  static constexpr timeType toTimeTypeUnitMulMax = 160000000 / second_th;
+  static constexpr timeType toTimeTypeUnitMul = F_CPU / second_th;
+
+  static timeType toTimeTypeUnit (const timeType userUnit) { return userUnit * toTimeTypeUnitMul; }
+  static timeType toUserUnit (const timeType cycles) { return cycles / toTimeTypeUnitMul; }
+  static constexpr timeType timeMax() { return std::numeric_limits<timeType>::max() / toTimeTypeUnitMulMax; }
+};
+
+struct TimeFastMillis: TimeFastCycles<1000>
+{
+};
+
+struct TimeFastMicros: TimeFastCycles<1000000>
+{
+};
+
+struct TimeFastNanos: TimeFastCyclesBase
+{
+  static constexpr timeType toTimeTypeUnitMulMax = 160000000 / 40000000; // 4
+  static constexpr timeType toTimeTypeUnitMul =        F_CPU / 40000000; // 2 or 4 (smallest possible)
+  static constexpr timeType toTimeTypeUnitDiv =                      25; // (mul/div=F_CPU/10^9)
+
+  static timeType toTimeTypeUnit (const timeType nanos) { return (nanos * toTimeTypeUnitMul) / toTimeTypeUnitDiv; }
+  static timeType toUserUnit (const timeType cycles) { return (cycles * toTimeTypeUnitDiv) / toTimeTypeUnitMul; }
+  static constexpr timeType timeMax() { return std::numeric_limits<timeType>::max() / toTimeTypeUnitMulMax; }
+};
+
+} //TimePolicy
 
-template <bool PeriodicT, typename YieldPolicyT = YieldPolicy::DoNothing>  
+template <bool PeriodicT, typename YieldPolicyT = YieldPolicy::DoNothing, typename TimePolicyT = TimePolicy::TimeMillis>
 class timeoutTemplate
 {
 public:
-  using timeType = decltype(millis());
-  
-  timeoutTemplate(timeType timeout) 
-    : _timeout(timeout), _start(millis())  
-  {} 
+  using timeType = typename TimePolicyT::timeType;
+
+  static constexpr timeType neverExpires = std::numeric_limits<timeType>::max();
+
+  timeoutTemplate(const timeType userTimeout)
+  {
+    reset(userTimeout);
+  }
 
   bool expired()
```

Comment:
If expiredRetrigger() and expiredOneShot() are in IRAM, then expired() and bool() probably should be as well.
BTW, why are any methods in IRAM? What is the use case for calling from an ISR?

Reply:
> why are any methods in IRAM? What is the use case for calling from an ISR?

It's about speed, not ISR requirement, so this comment remains true:
```
// "Fast" versions sacrifices time range for improved precision and reduced execution time (by 86%)
// (cpu cycles for ::expired(): 372 (millis()) vs 52 (ESP.getCycleCount()))
```